Ast/src/Lexer.cpp

```diff
@@ -535,37 +577,88 @@ Lexeme Lexer::readQuotedString()
             return Lexeme(Location(start, position()), Lexeme::BrokenString);
 
         case '\\':
+            readBackslashInString();
+            break;
+
+        default:
             consume();
-            switch (peekch())
-            {
-            case '\r':
-                consume();
-                if (peekch() == '\n')
-                    consume();
-                break;
+        }
+    }
 
-            case 0:
-                break;
+    consume();
 
-            case 'z':
-                consume();
-                while (isSpace(peekch()))
-                    consume();
-                break;
+    return Lexeme(Location(start, position()), Lexeme::QuotedString, &buffer[startOffset], offset - startOffset - 1);
+}
 
-            default:
-                consume();
-            }
+const Lexeme Lexer::nextInterpolatedString()
+{
+    Position start = position();
+    unsigned int startOffset = offset;
+
+    std::optional<Lexeme> readSectionOpt = readInterpolatedStringSection(start, Lexeme::InterpStringMid);
+
+    if (auto readSection = readSectionOpt)
+    {
+        lexeme = *readSection;
+        return lexeme;
+    }
+
+    consume();
+
+    lexeme = Lexeme(Location(start, position()), Lexeme::InterpStringEnd, &buffer[startOffset], offset - startOffset - 1);
+    return lexeme;
+}
+
+Lexeme Lexer::readInterpolatedStringBegin()
+{
+    Position start = position();
+    consume();
+
+    std::optional<Lexeme> readSectionOpt = readInterpolatedStringSection(start, Lexeme::InterpStringBegin);
+
+    if (!readSectionOpt)
+    {
+        return Lexeme(Location(start, position()), Lexeme::BrokenInterpNoFormat);
+    }
+
+    return *readSectionOpt;
+}
+
+std::optional<Lexeme> Lexer::readInterpolatedStringSection(Position start, Lexeme::Type formatType)
+{
+    unsigned int startOffset = offset;
+
+    while (peekch() != '`')
+    {
+        switch (peekch())
+        {
+        case 0:
+        case '\r':
+        case '\n':
+            return std::optional(Lexeme(Location(start, position()), Lexeme::BrokenString));
```

Comment:
I think you don't need an explicit call to optional ctor here and below, it's more idiomatic to simply return the value.